rfcs/0001-buildpacks-architecture.md

```diff
@@ -0,0 +1,130 @@
+
+This above architecture does not go well with the implementation CNB philosophy
+of "ask for your requirements at each stage, do one thing, and do it well".
+This leads to additional dependencies ending up in the final app image
+providing implicit rather than explicit behavior.
+
+Further, this structure makes it difficult to address complex issues like
+[#37](https://github.com/paketo-buildpacks/nodejs/issues/37)
+
+
+## Proposal
+
+Have 8 implementation CNBs:
+
+- node-engine:
+  - Provides `node` and `npm` (npm is the default package manager
+  that comes together with node) executable on PATH
+  - Requires none.
+
+- npm-install: installs dependencies to `node_modules` and provides them.
+  - Provides `node_modules`
+  - Requires `node` during `build`
+
+- yarn: provides `yarn` executable on PATH.
+  - Provides `yarn`
+  - Requires none.
+
+- yarn-install: installs dependencies to `node_modules` and provides them.
+  - Provides `node_modules`
+  - Requires {`node`, `yarn`} during `build`
+
+- tini: provides the [tini](https://github.com/krallin/tini) process manager,
+  which will be used in our start commands. This is needed to allow signals to
+  be passed to the node process spawned by the following start command CNBs
+  npm-start and yarn-start
+  - Provides `tini`
+  - Requires none.
+  See [issue #37](https://github.com/paketo-buildpacks/nodejs/issues/37) for motivation.
+
+
+- npm-start: sets up a start command that uses `tini` and `npm`
+  - Provides none
+  - Requires {`node`, `npm`, `node_modules`, `tini`} during `launch`
+  - Example start command would be `tini -g -- npm start`
+
+- yarn-start: sets up a start command that uses `tini` and `yarn`
+  - Provides none
+  - Requires {`node`, `yarn`, `node_modules`, `tini`} during `launch`
+
+- node-start: sets up a start command that uses `tini` and `node`
+  - Provides none
+  - Requires {`node`, `tini`} during `launch`
+  - Example start command would be `tini -g -- node server.js`
+
+The above implementation buildpacks should be structured as follows in the nodejs language family buildpack.
+
+```toml
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-engine"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn-install"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/tini"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn-start"
+    version = ""
+
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-engine"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/npm-install"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/tini"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/npm-start"
+    version = ""
+
+
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-engine"
+    version = ""
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-run"
```

Comment:
Is it `node-run` or `node-start` (which is what is listed earlier).  Also if y'all do intend to use `tini` here it needs to be added to the group.

Reply:
Original plan was to not include `tini` here. But open to suggestion.